charmcraft/commands/store/charmlibs.py

```diff
@@ -75,62 +67,69 @@ def create_charm_name_from_importable(charm_name: str) -> str:
 
 def get_lib_internals(lib_path: pathlib.Path) -> LibInternals:
     """Get content, its hash, and all the metadata fields from a library."""
-    metadata_fields = (b"LIBAPI", b"LIBPATCH", b"LIBID")
-    metadata = dict.fromkeys(metadata_fields)
-    hasher = hashlib.sha256()
-    with lib_path.open("rb") as fh:
-        for line in fh:
-            if line.startswith(metadata_fields):
-                try:
-                    field, value = [x.strip() for x in line.split(b"=")]
-                except ValueError:
-                    raise CraftError(
-                        "Bad metadata line in {!r}: {!r}".format(str(lib_path), line)
-                    ) from None
-                metadata[field] = value
-            else:
-                hasher.update(line)
-    content_hash = hasher.hexdigest()
-
-    missing = [k.decode("ascii") for k, v in metadata.items() if v is None]
+    content = lib_path.read_text()
+    try:
+        tree = ast.parse(content)
+    except Exception:
+        raise CraftError(f"Failed to parse Python library {str(lib_path)!r}")
+
+    def _api_patch_validator(value):
+        return isinstance(value, int) and value >= 0
+
+    FIELDS = {
+        "LIBAPI": (
+            _api_patch_validator,
+            "Library {!r} metadata field LIBAPI is not zero or a positive integer.",
+        ),
+        "LIBPATCH": (
+            _api_patch_validator,
+            "Library {!r} metadata field LIBPATCH is not zero or a positive integer.",
+        ),
+        "LIBID": (
+            lambda value: isinstance(value, str) and value and value.isascii(),
+            "Library {!r} metadata field LIBID must be a non-empty ASCII string.",
+        ),
+    }
+
+    # walk the AST "first layer", find assignments to the key fields and validate them
+    metadata = {}
+    for node in ast.iter_child_nodes(tree):
+        if isinstance(node, ast.Assign):
+            for target in node.targets:
+                if target.id in FIELDS:
+                    validator, error_msg = FIELDS[target.id]
+                    if not isinstance(node.value, ast.Constant):
+                        raise CraftError(error_msg.format(str(lib_path)))
```

Comment:
Is this something we might want to consider a different message for? E.g. if a user thinks they're being very smart and makes the file say:

```python
LIBAPI = int(open('VERSION').read().split('.')[0])
```

we might want to let them know that the message needs to be a constant rather than just evaluating to an `int >= 0`.

Reply:
Nice idea.

Reply:
Actually, it was more complicated that sounded :p 

The problem is when the user does `LIBPATCH = -1` (which may be more probable than a twisted expression as you used), and that ends up in an error like "you need to put a constant". That's because `-1` is actually an unary operation...

So I just improved the message to be explicit about needing a constant, and it should work.

Thanks!!!